### ISSUE TITLE
ext_authz not passing content-type in POST request

### ISSUE THREAD

**Issue by digressionary** [envoyproxy/envoy#7338]
**Issue Template**

*Title*: *ext_authz not passing content-type in POST request*

*Description*:
>I POST a request through Ambassador API gateway which directs it to a Node.JS app.  I am passing a body in the request like this:

```curl -X POST \
  http://localhost/example/ \
  -H 'Content-Type: application/json' \
  -H 'Postman-Token: a00e1ecc-c567-4df1-94ee-c2d39c6618e8' \
  -H 'cache-control: no-cache' \
  -d '{"username":"redacted","password":"redacted"}'
```

I have sent the request through this echo service; https://github.com/jmalloc/echo-server
to see what request comes through Ambassador.

My ecosystem:
  node: 10.16.0
  express: 4.17.1
  body-parser: 1.19.0
  concat-stream: 2.0.0
  Docker: 19.03
  Kubernetes: 1.14.1
  Ambassador API Gateway: 0.72.0

My Ambassador configuration:
```apiVersion: v1
kind: Service
metadata:
  name: {{ template "fullname" . }}
  labels:
    app: {{ template "name" . }}
    chart: {{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}
    release: {{ .Release.Name }}
    heritage: {{ .Release.Service }}
  annotations:
    getambassador.io/config: |
      ---
      apiVersion: ambassador/v1
      kind:  AuthService
      name:  authentication
      auth_service: {{ template "fullname" . }}.{{ .Release.Namespace }}:{{ .Values.service.externalPort }}
      proto: http
      allowed_request_headers:
        - requested-status
        - requested-body-info
        - content-type
      allowed_authorization_headers:
        - auth-request-body
      include_body:
        max_bytes: 4096
        allow_partial: true
      ---
      apiVersion: ambassador/v1
      kind:  Mapping
      name: echo-service
      prefix: /
      service: echo-service.default:80
      host_rewrite: ""
```

Here is what the echo service sees:
```
Request served by echo-deployment-b6b6878c-7vmnh

HTTP/1.1 POST /example/

Host: localhost
Content-Type: application/json
Postman-Token: 7797f187-6676-4a32-bb14-0e6cb9aeac3a
Accept: */*
User-Agent: PostmanRuntime/7.6.0
Accept-Encoding: gzip, deflate
Content-Length: 45
X-Request-Id: 233bb9de-cdc4-41a4-a54b-d2a3b28dcb4e
X-Envoy-Expected-Rq-Timeout-Ms: 3000
X-Envoy-Original-Path: /example/
Cache-Control: no-cache
X-Forwarded-For: 192.168.65.3
X-Forwarded-Proto: http
X-Envoy-Internal: true

{"username":"redacted","password":"redacted"}
```

However, if I use this express app:
https://gist.githubusercontent.com/digressionary/86c48ccf89fc210d05ae5e76d8390b7d/raw/fa66d729927ce594ee146cf444b168eec170fc32/index.js

When I console.log out req.body, it is an empty object {}.

According to the documenation on the express framework, the body-parser will fail if no content type is set.

The STDERR for the body parser points to the problem:
Thu, 20 Jun 2019 15:05:09 GMT body-parser:urlencoded content-type undefined
Thu, 20 Jun 2019 15:05:09 GMT body-parser:urlencoded skip parsing
Thu, 20 Jun 2019 15:05:09 GMT body-parser:json content-type undefined
Thu, 20 Jun 2019 15:05:09 GMT body-parser:json skip parsing

Request headers according to the express app:
```{ host: 'localhost',
  'content-length': '45',
  'x-forwarded-proto': 'http',
  'x-forwarded-for': '192.168.65.3,10.1.0.115',
  'user-agent': 'PostmanRuntime/7.6.0',
  'x-envoy-internal': 'true',
  'x-envoy-expected-rq-timeout-ms': '5000' }
```

The expected behaviour would be for the express app to receive a content-type header as passed in the request.

**Comment by mattklein123** [comment]
cc @gsagula @dio

**Comment by gsagula** [comment]
@dio I think I know what is going on here. I will catch up with some work
tomorrow so I can take a look at this issue too.
-- 
Gabriel Linden Sagula

**Comment by dio** [comment]
@gsagula OK. Not sure what corresponding Ambassador version of Envoy here. But from my quick check [here](https://blog.getambassador.io/ambassador-0-72-ddaca31c75f8), seems like it is 1.10 that doesn't have: `allowed_request_headers:` and `allowed_authorization_headers:` or is it translated directly to: 

```
                authorization_request:
                  allowed_headers:
                    patterns:
```
?

```yaml
          - name: envoy.ext_authz
            config:
              http_service:
                path_prefix:
                server_uri:
                  uri: http://dump:8000
                  cluster: service_authz
                  timeout: 5s
                authorization_request:
                  allowed_headers:
                    patterns:
                    - exact: content-type
```

I did a repro using above config, and dump the headers from the "dump" service I get:

```
dump_1   | [03:02:41] POST {
dump_1   |   host: '192.168.99.100:10000',
dump_1   |   'content-length': '12',
dump_1   |   'content-type': 'application/json',
dump_1   |   'x-envoy-internal': 'true',
dump_1   |   'x-forwarded-for': '172.30.0.3',
dump_1   |   'x-envoy-expected-rq-timeout-ms': '5000'
dump_1   | } body={"ok": true}
```

For a request:

```
curl -X POST -H "Content-Type: application/json" -d '{"ok": true}' $(docker-machine ip):10000
```

**Comment by gsagula** [comment]
@digressionary I'm also having a hard time trying to reproduce the issue. Here is my config:
```
---
apiVersion: ambassador/v1
kind: Module
name: ambassador
config: {}
---
apiVersion: ambassador/v1
kind: AuthService
name: authentication
auth_service: "auth-service:8080"
path_prefix: "/extauth"
proto: http
allowed_request_headers:
- requested-status
- content-type
allowed_authorization_headers:
- auth-request-body
include_body:
  max_bytes: 2048
  allow_partial: true
---
apiVersion: ambassador/v1
kind:  Mapping
name: echo-service
prefix: /
service: http://httpbin.org
host_rewrite: httpbin.org
```

curl -X POST -d "{ \"test\": 123 }" -H "Content-Type: application/json" -H "requested-status:200" http://localhost:61880/post
```
{
  "args": {},
  "data": "{ \"test\": 123 }",
  "files": {},
  "form": {},
  "headers": {
    "Accept": "*/*",
    "Auth-Request-Body": "{ \"test\": 123 }",
    "Content-Length": "15",
    "Content-Type": "application/json",
    "Host": "httpbin.org",
    "Requested-Status": "200",
    "User-Agent": "curl/7.54.0",
    "X-B3-Parentspanid": "92636a73df407aa4",
    "X-B3-Sampled": "1",
    "X-B3-Spanid": "3592b5d2775da840",
    "X-B3-Traceid": "92636a73df407aa4",
    "X-Envoy-Expected-Rq-Timeout-Ms": "3000",
    "X-Envoy-Internal": "true",
    "X-Envoy-Original-Path": "/post"
  },
  "json": {
    "test": 123
  },
  "origin": "172.19.0.1, 216.71.203.120, 172.19.0.1",
  "url": "https://httpbin.org/post"
}
```

@dio It should translate it to the correct Envoy config:
```
---
apiVersion: ambassador/v1
kind: AuthService
name: authentication
auth_service: "auth-service:8080"
path_prefix: "/extauth"
proto: http
allowed_request_headers:
- requested-status
- content-type
allowed_authorization_headers:
- auth-request-body
include_body:
  max_bytes: 4096
  allow_partial: true
```
will give Envoy something like this:
```
{
                                                "http_service": {
                                                    "authorization_request": {
                                                        "allowed_headers": {
                                                            "patterns": [
                                                                {
                                                                    "exact": "proxy-authorization"
                                                                },
                                                                {
                                                                    "exact": "user-agent"
                                                                },
                                                                {
                                                                    "exact": "x-forwarded-host"
                                                                },
                                                                {
                                                                    "exact": "from"
                                                                },
                                                                {
                                                                    "exact": "cookie"
                                                                },
                                                                {
                                                                    "exact": "authorization"
                                                                },
                                                                {
                                                                    "exact": "x-forwarded-for"
                                                                },
                                                                {
                                                                    "exact": "content-type"
                                                                },
                                                                {
                                                                    "exact": "requested-status"
                                                                },
                                                                {
                                                                    "exact": "x-forwarded-proto"
                                                                }
                                                            ]
                                                        }
                                                    },
                                                    "authorization_response": {
                                                        "allowed_client_headers": {
                                                            "patterns": [
                                                                {
                                                                    "exact": "location"
                                                                },
                                                                {
                                                                    "exact": "auth-request-body"
                                                                },
                                                                {
                                                                    "exact": "set-cookie"
                                                                },
                                                                {
                                                                    "exact": "proxy-authenticate"
                                                                },
                                                                {
                                                                    "exact": "www-authenticate"
                                                                },
                                                                {
                                                                    "exact": "authorization"
                                                                }
                                                            ]
                                                        },
                                                        "allowed_upstream_headers": {
                                                            "patterns": [
                                                                {
                                                                    "exact": "location"
                                                                },
                                                                {
                                                                    "exact": "auth-request-body"
                                                                },
                                                                {
                                                                    "exact": "set-cookie"
                                                                },
                                                                {
                                                                    "exact": "proxy-authenticate"
                                                                },
                                                                {
                                                                    "exact": "www-authenticate"
                                                                },
                                                                {
                                                                    "exact": "authorization"
                                                                }
                                                            ]
                                                        }
                                                    },
                                                    "path_prefix": "/extauth",
                                                    "server_uri": {
                                                        "cluster": "cluster_extauth_auth_service_8080",
                                                        "timeout": "5.000s",
                                                        "uri": "http://extauth"
                                                    }
                                                },
                                                "with_request_body": {
                                                    "allow_partial_message": true,
                                                    "max_request_bytes": 4096
                                                }
                                            },
                                            "name": "envoy.ext_authz"
                                        }
```

**Comment by digressionary** [comment]
I realize I made a mistake in my example.
This is the list of headers I originally whitelisted:
allowed_request_headers:
- requested-status

However, later through testing, I found that this worked:
allowed_request_headers:
- requested-status
- content-type

In older versions of Ambassador, this was the default behaviour.  Maybe documentation clarification?  Also, there's no mention in the documentation of headers requested-status, requested-body-info, could we clarify these also in the docs?

I will close the bug because whitelisting content-type did fix the problem.  So it is not a bug, more of a lack of clarity.

Best,
